Spec/RealtimeClient.swift

```diff
@@ -352,6 +352,478 @@ class RealtimeClient: QuickSpec {
+                                completion(getTestTokenDetails(clientId: "tester"), nil)
+                            }
+
+                            client.auth.authorize(tokenParams, options: authOptions) { tokenDetails, error in
+                                guard let error = error else {
+                                    fail("ErrorInfo is nil"); done(); return
+                                }
+                                expect(error.code).to(equal(40102))
+                                expect(error.description).to(contain("incompatible credentials"))
+                                expect(tokenDetails).to(beNil())
+
+                                expect(client.connection.state).to(equal(ARTRealtimeConnectionState.Connecting))
+
+                                guard let transport = client.transport as? TestProxyTransport else {
+                                    fail("TestProxyTransport is not set"); done(); return
+                                }
+                                expect(transport.protocolMessagesReceived.filter({ $0.action == .Connected })).to(haveCount(0))
+                                expect(transport.protocolMessagesReceived.filter({ $0.action == .Error })).to(haveCount(0))
+                                done()
+                            }
+                        }
+                        client.connect()
+                    }
+                }
+
+                // RTC8c
+                it("when the connection is in the DISCONNECTED, SUSPENDED, FAILED, or CLOSED state when auth#authorize is called, after obtaining a token the library should move to the CONNECTING state and initiate a connection attempt using the new token") {
+                    let options = AblyTests.commonAppSetup()
+                    options.autoConnect = false
+                    let testToken = getTestToken()
+                    options.token = testToken
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+                    client.setTransportClass(TestProxyTransport.self)
+
+                    waitUntil(timeout: testTimeout) { done in
+                        client.connection.once(.Connected) { stateChange in
+                            expect(stateChange?.reason).to(beNil())
+                            done()
+                        }
+                        client.connect()
+                    }
+
+                    waitUntil(timeout: testTimeout) { done in
+                        let partialDone = AblyTests.splitDone(3, done: done)
+
+                        client.connection.once(.Connecting) { stateChange in
+                            guard let stateChange = stateChange else {
+                                fail("ConnectionStateChange is nil"); partialDone(); return
+                            }
+                            expect(stateChange.previous).to(equal(ARTRealtimeConnectionState.Closed))
```

Comment:
I appreciate you are checking that the connection is `CLOSED`, but you are not making it `CLOSED` before calling `authorize` so there is a race condition here.